Client/Coordinators/ShareExtensionCoordinator.swift

```diff
@@ -0,0 +1,150 @@
+            dequeueNotShownJSAlert()
+        default:
+            dequeueNotShownJSAlert()
+        }
+    }
+
+    private func showSendToDevice() {
+        guard let selectedTab = tabManager.selectedTab,
+              let url = selectedTab.canonicalURL?.displayURL
+        else { return }
+
+        let themeColors = themeManager.currentTheme.colors
+        let colors = SendToDeviceHelper.Colors(defaultBackground: themeColors.layer1,
+                                               textColor: themeColors.textPrimary,
+                                               iconColor: themeColors.iconPrimary)
+        let shareItem = ShareItem(url: url.absoluteString,
+                                  title: selectedTab.title)
+
+        let helper = SendToDeviceHelper(
+            shareItem: shareItem,
+            profile: profile,
+            colors: colors,
+            delegate: self)
+        let viewController = helper.initialViewController()
+
+        TelemetryWrapper.recordEvent(category: .action, method: .tap, object: .sendToDevice)
+        router.present(viewController)
+    }
+
+    private func dequeueNotShownJSAlert() {
+        guard let alertInfo = tabManager.selectedTab?.dequeueJavascriptAlertPrompt()
+        else {
+            parentCoordinator?.didFinish(from: self)
+            return
+        }
+        let alertController = alertInfo.alertController()
+        alertController.delegate = self
+        router.present(alertController)
+    }
+}
+
+extension ShareExtensionCoordinator: DevicePickerViewControllerDelegate, InstructionsViewDelegate {
+    func devicePickerViewControllerDidCancel(_ devicePickerViewController: DevicePickerViewController) {
+        router.dismiss()
+        parentCoordinator?.didFinish(from: self)
+    }
+
+    func devicePickerViewController(_ devicePickerViewController: DevicePickerViewController, didPickDevices devices: [RemoteDevice]) {
+        guard let shareItem = devicePickerViewController.shareItem
+        else {
+            router.dismiss()
+            parentCoordinator?.didFinish(from: self)
+            return
+        }
+
+        guard shareItem.isShareable else {
+            let alert = UIAlertController(title: .SendToErrorTitle, message: .SendToErrorMessage, preferredStyle: .alert)
+            alert.addAction(UIAlertAction(title: .SendToErrorOKButton, style: .default) { [weak self] _ in self?.router.dismiss() })
+            router.present(alert, animated: true) { [weak self] in
+                guard let self = self else { return }
+                self.parentCoordinator?.didFinish(from: self)
+            }
+            return
+        }
+        profile.sendItem(shareItem, toDevices: devices).uponQueue(.main) { [weak self] _ in
+            guard let self = self else { return }
+            self.router.dismiss()
+            self.parentCoordinator?.didFinish(from: self)
+            DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
```

Comment:
I guess we still need this `asyncAfter`, but would be good to confirm when everything is piped together

Reply:
ok perfect.